### ISSUE TITLE
Support arm64 devices

### ISSUE THREAD

**Issue by jontze** [jontze/cadency-rs#94]
Currently there's only a single docker image for amd64 devices available but not for arm64. To give users the opportunity to run the docker image and cadency on arm64 devices we have to adjust the image building process. It might also be required to refactor some parts of cadency to be able to run on arm64.

I'm not that familiar with cross compiling in rust and docker multi-platform builds, so help and contributions are highly appreciated.

Came up during #90 

**Implementation hints:**

- [cross](https://github.com/cross-rs/cross) is a nice tool to test and build on multiple platforms

**Comment by jontze** [comment]
I adjusted the image building workflow in the CI to create the image for multiple selected platforms. Unfortunately, the `cargo build` process fails on `linux/arm64`. I have to take a look at this and check how I can investigate what's causing this - on my amd64 machine :laughing:

**Comment by TeaDrinkingProgrammer** [comment]
> I adjusted the image building workflow in the CI to create the image for multiple selected platforms. Unfortunately, the `cargo build` process fails on `linux/arm64`. I have to take a look at this and check how I can investigate what's causing this - on my amd64 machine 😆

I might be able to help, as I have a Raspberry Pi 3b

**Comment by jontze** [comment]
> I might be able to help, as I have a Raspberry Pi 3b

Nice, that would be a great help! As I could successfully cross-compile an executable with [cross](https://github.com/cross-rs/cross) for arm64, I have the assumption that in the cadency dockerfile just some build dependencies might be missing on arm64.

If you clone the repo to the pi and switch to the `develop` branch, can you successfully run:
- `cargo build` or
- `docker build . -t <some_local_image_name>` ?
And if not, what error do you get?

**Comment by TeaDrinkingProgrammer** [comment]
I had several issues:
The first was with GCC, which I had to install.
The second was with https://lib.rs/crates/audiopus_sys. I installed opus to solve it, but to cross compile, it probably needs to be statically linked as described in the docs.

Then I got this error, but I honestly don't know what happens there, except that it is a problem with Serenity. It takes ages to compile it, and then it gives this error.

```
error: could not compile `serenity`

Caused by:
  process didn't exit successfully: `rustc --crate-name serenity --edition=2018 /home/rocky/.cargo/registry/src/github.com-1ecc6299db9ec823/serenity-0.11.5/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 --cfg 'feature="async-tungstenite"' --cfg 'feature="base64"' --cfg 'feature="builder"' --cfg 'feature="bytes"' --cfg 'feature="cache"' --cfg 'feature="client"' --cfg 'feature="dashmap"' --cfg 'feature="flate2"' --cfg 'feature="gateway"' --cfg 'feature="http"' --cfg 'feature="model"' --cfg 'feature="parking_lot"' --cfg 'feature="reqwest"' --cfg 'feature="rustls_backend"' --cfg 'feature="tokio"' --cfg 'feature="typemap_rev"' --cfg 'feature="utils"' --cfg 'feature="voice"' -C metadata=15c09442d83ef4ac -C extra-filename=-15c09442d83ef4ac --out-dir /home/rocky/cadency-rs/target/debug/deps -L dependency=/home/rocky/cadency-rs/target/debug/deps --extern async_trait=/home/rocky/cadency-rs/target/debug/deps/libasync_trait-b38d7ee67981a03b.so --extern async_tungstenite=/home/rocky/cadency-rs/target/debug/deps/libasync_tungstenite-204034cdfeba29b9.rmeta --extern base64=/home/rocky/cadency-rs/target/debug/deps/libbase64-50e421855f229ec5.rmeta --extern bitflags=/home/rocky/cadency-rs/target/debug/deps/libbitflags-b53b2de0be2743ed.rmeta --extern bytes=/home/rocky/cadency-rs/target/debug/deps/libbytes-58ad704c6d97ad6a.rmeta --extern cfg_if=/home/rocky/cadency-rs/target/debug/deps/libcfg_if-15daa030cca31511.rmeta --extern dashmap=/home/rocky/cadency-rs/target/debug/deps/libdashmap-46ce3279607629df.rmeta --extern flate2=/home/rocky/cadency-rs/target/debug/deps/libflate2-01fee8db2757dea6.rmeta --extern futures=/home/rocky/cadency-rs/target/debug/deps/libfutures-d8e08b92215c5494.rmeta --extern mime=/home/rocky/cadency-rs/target/debug/deps/libmime-ffe4c2949446184c.rmeta --extern mime_guess=/home/rocky/cadency-rs/target/debug/deps/libmime_guess-f8971f2631210377.rmeta --extern parking_lot=/home/rocky/cadency-rs/target/debug/deps/libparking_lot-bee0770e446193ef.rmeta --extern percent_encoding=/home/rocky/cadency-rs/target/debug/deps/libpercent_encoding-432e007c255cdd1c.rmeta --extern reqwest=/home/rocky/cadency-rs/target/debug/deps/libreqwest-9b8ac50fe01776a8.rmeta --extern serde=/home/rocky/cadency-rs/target/debug/deps/libserde-d24991e7e0d43d5d.rmeta --extern serde_value=/home/rocky/cadency-rs/target/debug/deps/libserde_value-f15511e8d4a00be2.rmeta --extern serde_json=/home/rocky/cadency-rs/target/debug/deps/libserde_json-ebf9c1e012c70a89.rmeta --extern dep_time=/home/rocky/cadency-rs/target/debug/deps/libtime-a5f95da331549399.rmeta --extern tokio=/home/rocky/cadency-rs/target/debug/deps/libtokio-d63465359bf3fbd6.rmeta --extern tracing=/home/rocky/cadency-rs/target/debug/deps/libtracing-e2376eaff3e4a960.rmeta --extern typemap_rev=/home/rocky/cadency-rs/target/debug/deps/libtypemap_rev-43a236e5474921de.rmeta --extern url=/home/rocky/cadency-rs/target/debug/deps/liburl-5bc4eaebbeee03f6.rmeta --cap-lints allow -L native=/home/rocky/cadency-rs/target/debug/build/ring-587a5bf5b155afc4/out` (signal: 9, SIGKILL: kill)
```

**Comment by jontze** [comment]
Thanks for trying this out and for looking into this. 

The error message sadly isn't giving that much information about what actually went wrong :thinking: 

Before I'll reach out to the serenity guys, a few thinks that I could find out after reading through the serenity docs and similar errors in other projects:
- Try if this error persists after running `cargo clean` and then build again - or in docker to build without cache should do the job as well (removes potential corrupted build artifacts) 
- Which rust version are you on? Serenity's defacto MSRV is 1.59 (https://github.com/serenity-rs/serenity#msrv-policy)
- How much memory does you PI has? And could you monitor the memory usage during the compiling? I read that such a sigkill is almost always caused by the system running out of memory. (e.g.: https://github.com/rust-lang/rust/issues/51309 )

**Comment by TeaDrinkingProgrammer** [comment]
- I tried cargo clean, doesn't change anything
- Rustc --version returns 1.67.1
- The Pi has 500mb of RAM, so that is probably the issue 😄

**Comment by jontze** [comment]
I finally had the chance to test the docker build locally on a Raspberry Pi Modell B with 8GBs of RAM and it worked perfectly fine (beside of that it took nearly half an hour) :laughing: .

I found out that if I set an environment variable to adjust the way how cargo fetches the dependencies on during the arm64 build in the CI (https://github.com/rust-lang/cargo/issues/10230#issuecomment-1001349815), the building seem to work in the CI as well. However, the build is way slower, but it's working for now. I'll look further into this. :thinking:
